# An Important Note
I worked on this project during my studies for Dataquest online Data Science Bootcamp. This was for "Python for Data Science: Intermediate" part of the bootcamp.

# Exploring Hackers News Posts
In this project, I will work on a data set for "Hackers News Posts". Hackers News is a very popular website where users submit posts usually about technology in two different types.<br/>
1. Ask HN = This is one of the types of posts which users ask special questions to get answers from different people.
2. Show HN = This is the other type of the posts which users share their projects,products or anything which is interesting.<br/>

<br/>
I'll compare these two types of posts to determine the following:<br/>
1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on average?

I will start by importing the libraries we need and reading the data set into a list of lists.

In [1]:
# Reading in the data
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

In [2]:
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


# Removing The Names of The Columns
Now I will separate the first row which has the names of the columns from the data set and assign it to "headers". After that I will assign the rest of the data set back to "hn".

In [3]:
# Removing the names of the columns
headers = hn[0]
hn = hn[1:]

In [4]:
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


As it can be seen, the data set has different types of informations such as titles of the posts, time of creating for each of the posts, number of the comments for each of the posts.

# Identification of Ask HN and Show HN Posts
Before starting analyzing the data set, I want to separate it into two different parts for two different types of the posts. This will make analyzing easier for me.

In [5]:
# Identify and separate the Ask HN and Show HN posts.
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn') == True:
        ask_posts.append(row)
    elif title.startswith('show hn') == True:
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


# Calculating the Average Number of Comments for Ask HN and Show HN Posts
Now that I separated ask posts and show posts into different lists, I'll calculate the average number of comments each type of post receives.

In [6]:
# Calculating the average number of comments for Ask HN posts.
total_ask_comments = 0
for row in ask_posts:
    num_comments_ask = int(row[4])
    total_ask_comments += num_comments_ask
    avg_ask_comments = total_ask_comments / len(ask_posts)
print(total_ask_comments)
print(len(ask_posts))
print(avg_ask_comments)

24483
1744
14.038417431192661


As it can be seen, there were total of 24483 comments for 1744 pieces of Ask HN posts and the average number of comments is 14.04.

In [7]:
total_show_comments = 0
for row in show_posts:
    num_comments_show = int(row[4])
    total_show_comments += num_comments_show
    avg_show_comments = total_show_comments / len(show_posts)
print(total_show_comments)
print(len(show_posts))
print(avg_show_comments)

11988
1162
10.31669535283993


As it can be seen, there were total of 11988 comments for 1162 pieces of Show HN posts and the average number of comments is 10.32.

# Finding the Amount of Ask HN Posts and Comments by Hour Created
Since ask posts are more likely to receive comments, I'll focus my remaining analysis just on these posts.<br/>
Next, I'll determine if I can maximize the amount of comments an Ask HN post receives by creating it at a certain time. First, I'll find the amount of Ask HN posts created during each hour of day, along with the number of comments those posts received. Then, I'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [8]:
import datetime as dt

In [9]:
# Calculating the amount of ask posts created during each hour of day and the number of comments received.
result_list = []
for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    at_com = [created_at, num_comments]
    result_list.append(at_com)
print(result_list[:5])


[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [10]:
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    time = row[0]
    time_dt = dt.datetime.strptime(time, "%m/%d/%Y %H:%M")
    hour = time_dt.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = 1
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]

In [11]:
print(counts_by_hour)

{'20': 80, '22': 71, '00': 55, '15': 116, '16': 108, '06': 44, '01': 60, '21': 109, '14': 107, '10': 59, '19': 110, '18': 109, '05': 46, '04': 47, '02': 58, '07': 34, '23': 68, '12': 73, '11': 58, '09': 45, '17': 100, '13': 85, '03': 54, '08': 48}


In [12]:
print(comments_by_hour)

{'20': 1721, '22': 478, '00': 438, '15': 4477, '16': 1798, '06': 397, '01': 651, '21': 1742, '14': 1414, '10': 793, '19': 1186, '18': 1438, '05': 436, '04': 335, '02': 1379, '07': 266, '23': 543, '12': 684, '11': 640, '09': 246, '17': 1146, '13': 1225, '03': 421, '08': 488}


# Calculating the Average Number of Comments for Ask HN Posts by Hour

In [13]:
# Calculating the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []
for hour in comments_by_hour:
    avg_by_hour.append([hour, (comments_by_hour[hour] / counts_by_hour[hour])])
avg_by_hour

[['20', 21.5125],
 ['22', 6.732394366197183],
 ['00', 7.963636363636364],
 ['15', 38.5948275862069],
 ['16', 16.64814814814815],
 ['06', 9.022727272727273],
 ['01', 10.85],
 ['21', 15.98165137614679],
 ['14', 13.214953271028037],
 ['10', 13.440677966101696],
 ['19', 10.781818181818181],
 ['18', 13.192660550458715],
 ['05', 9.478260869565217],
 ['04', 7.127659574468085],
 ['02', 23.775862068965516],
 ['07', 7.823529411764706],
 ['23', 7.985294117647059],
 ['12', 9.36986301369863],
 ['11', 11.03448275862069],
 ['09', 5.466666666666667],
 ['17', 11.46],
 ['13', 14.411764705882353],
 ['03', 7.796296296296297],
 ['08', 10.166666666666666]]

# Sorting and Printing Values from a List of Lists

In [14]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)

[[21.5125, '20'], [6.732394366197183, '22'], [7.963636363636364, '00'], [38.5948275862069, '15'], [16.64814814814815, '16'], [9.022727272727273, '06'], [10.85, '01'], [15.98165137614679, '21'], [13.214953271028037, '14'], [13.440677966101696, '10'], [10.781818181818181, '19'], [13.192660550458715, '18'], [9.478260869565217, '05'], [7.127659574468085, '04'], [23.775862068965516, '02'], [7.823529411764706, '07'], [7.985294117647059, '23'], [9.36986301369863, '12'], [11.03448275862069, '11'], [5.466666666666667, '09'], [11.46, '17'], [14.411764705882353, '13'], [7.796296296296297, '03'], [10.166666666666666, '08']]


In [15]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap[:5]

[[38.5948275862069, '15'],
 [23.775862068965516, '02'],
 [21.5125, '20'],
 [16.64814814814815, '16'],
 [15.98165137614679, '21']]

In [16]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.78 average comments per post
20:00: 21.51 average comments per post
16:00: 16.65 average comments per post
21:00: 15.98 average comments per post


The time that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

# Conclusion
In this project, I analyzed "Ask HN" posts and "Show HN" posts to determine which type of post and time receive the most comments on average. Based on my analysis, to maximize the amount of comments a post receives, I'd recommend the post be categorized as "Ask HN" post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).